### PR TITLE
Feat/relations in es

### PR DESCRIPTION
**Pull request by suraj5077**
## Publishing relationship events to Kafka for syncing in ES.

Atlas will publish Relationship CUD events based on relationship types. Notifications enabled on new Kafka topic.

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

### Development

- [x] Lint rules pass locally
- [x] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### Code review 

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [x] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/JanusUtils.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.apache.atlas.repository.graphdb.janus;
+
+import org.janusgraph.util.encoding.LongEncoding;
+
+import javax.annotation.Nonnull;
+import java.util.Objects;
+
+public class JanusUtils {
```
Comment (n5nk) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/JanusUtils.java:8]: add final to the class , and also add a private constructor.

File: intg/src/main/java/org/apache/atlas/AtlasConfiguration.java [hunk 2]
```diff
@@ -91,7 +93,8 @@ public enum AtlasConfiguration {
     GRAPH_TRAVERSAL_PARALLELISM("atlas.graph.traverse.bucket.size",10),
     LINEAGE_ON_DEMAND_ENABLED("atlas.lineage.on.demand.enabled", true),
     LINEAGE_ON_DEMAND_DEFAULT_NODE_COUNT("atlas.lineage.on.demand.default.node.count", 3),
-    LINEAGE_MAX_NODE_COUNT("atlas.lineage.max.node.count", 9000);
+    LINEAGE_MAX_NODE_COUNT("atlas.lineage.max.node.count", 9000),
+    SUPPORTED_RELATIONSHIP_EVENTS("atlas.notification.relationships.filter", "asset_readme,asset_links,RelatedReadme,AtlasGlossarySemanticAssignment,AtlasGlossarySynonym,AtlasGlossaryAntonym");
```
Comment (n5nk) [inline intg/src/main/java/org/apache/atlas/AtlasConfiguration.java]: you can update the list with `readme` for now.

File: intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationship.java [hunk 3]
```diff
@@ -235,6 +241,11 @@ public AtlasRelationship(Map map) {
                     }
                 }
             }
+
+            if (MapUtils.isNotEmpty((Map) customRelationshipInfo) ) {
```
Comment (n5nk) [inline intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationship.java:245]: is `customRelationshipInfo` always `non-null` ?

File: intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationshipHeader.java [hunk 4]
```diff
@@ -51,6 +52,8 @@ public class AtlasRelationshipHeader extends AtlasStruct implements Serializable
     private String                                  label               = null;
     private AtlasObjectId                           end1                = null;
     private AtlasObjectId                           end2                = null;
+    private Map<String, Object> customRelationshipInfo;
```
Comment (n5nk) [inline intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationshipHeader.java:55]: nit: Can you check `customRelationshipInfo` requires any initialization ?

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasRelationshipStoreV2.java [hunk 5]
```diff
@@ -408,28 +424,33 @@ public AtlasRelationship getOrCreate(AtlasRelationship relationship) throws Atla
 
         // check if relationship exists
         AtlasEdge relationshipEdge = getRelationship(end1Vertex, end2Vertex, relationship);
-
+        boolean isCreated = false;
```
Comment (n5nk) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasRelationshipStoreV2.java:427]: what is `isCreated` referring to ? can you rename to be specific

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java [hunk 6]
```diff
@@ -2285,4 +2281,28 @@ public void repairHasLineageWithAtlasEdges(Set<AtlasEdge> inputOutputEdges) {
...
+        }
+    }
+
+    private void recordOutGoingEdgesToBeDeleted(Iterable<AtlasEdge> outgoingEdges) throws AtlasBaseException {
+        for (AtlasEdge edge : outgoingEdges) {
+            if (isRelationshipEdge(edge))
+                AtlasRelationshipStoreV2.recordRelationshipMutation(AtlasRelationshipStoreV2.RelationshipMutation.RELATIONSHIP_HARD_DELETE, edge, entityRetriever);
+        }
+    }
+
```
Comment (n5nk) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java:2305]: nit: remove extra lines.

File: webapp/src/main/java/org/apache/atlas/web/rest/RelationshipREST.java [hunk 7]
```diff
@@ -190,8 +186,7 @@ public void deleteByIds(List<String> guids) throws AtlasBaseException {
             if (AtlasPerfTracer.isPerfTraceEnabled(PERF_LOG)) {
                 perf = AtlasPerfTracer.getPerfTracer(PERF_LOG, "RelationshipREST.deleteById(" + guids.size() + ")");
             }
-
-            relationshipStore.deleteByIds(guids);
+            relationshipStore.deleteByIds(guids, true);
```
Comment (n5nk) [inline webapp/src/main/java/org/apache/atlas/web/rest/RelationshipREST.java:189]: @suraj5077  I see we pass `true` for all actions to notify.. Why it cant be at single place, while notification, check if notifying is enabled or not..

Comment (suraj5077) [reply webapp/src/main/java/org/apache/atlas/web/rest/RelationshipREST.java:189]: This is a different flow @n5nk . Notification check is enabled only at a single place. For that check sendNotifications method in class AtlasRelationshipStoreV2.

File: intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationshipHeader.java (hunk 4 above)
Comment (suraj5077) [reply intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationshipHeader.java:55]: not required

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasRelationshipStoreV2.java (hunk 5 above)
Comment (suraj5077) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasRelationshipStoreV2.java:427]: old PR

**Pull request merged**